### DISCUSSION TITLE
Can't bind to DirectProperty made from struct.

### DISCUSSION THREAD

**Discussion by sonicsmooth** [AvaloniaUI/Avalonia#10248] [General]
Edit -- answer is that property backing field needs to be non-nullable and `Parse` needs to be implemented.

What constraints must a DirectProperty made from a custom struct satisfy so I can bind to it in xaml?  Looks like I nee `Parse` method and custom type must not be nullable, right?

In my custom `TemplatedControl` I have an `Avalonia.DirectProperty` which is an instance of a custom struct `Distance` and is named `Value`.

    public class DistanceBox : TemplatedControl
    {
        public static readonly DirectProperty<DistanceBox, Distance?> ValueProperty =
            AvaloniaProperty.RegisterDirect<DistanceBox, Distance?>(
                nameof(Value),
                o => o.Value,
                (o, v) => o.Value = v);
        private Distance? _value = Distance.Millimeter;
        private Distance? Value
        {
            get => _value;
            set => this.SetAndRaise(ValueProperty, ref _value, value);
        }
        ...
    }

This `Distance`  struct can take a `string` as a constructor, so I can say `new Distance("15 mm")`, and the `ToString()` is implemented reasonably.  So I would expect this to work in xaml, but I cannot bind to or set this `Value` property in the xaml.  It doesn't come up in Intellisense, even though other properties using primitives do come up, such as `Title` and `SelectedUnitIndex` in the below example, which are just `string` and `int`, respectively.  Making this a class instead of a struct doesn't help.  Do I need a value converter?

(edit -- I discovered `Parse` allows Intellisense recognition but causes crash at runtime, complaining `Unhandled exception. System.MethodAccessException: Attempt by method 'avgrid.Views.MainWindow.!XamlIlPopulate(System.IServiceProvider, avgrid.Views.MainWindow)' to access method 'avgrid.Controls.DistanceBox.set_Value(System.Nullable``1<avgrid.Distances.Distance>)'`)

![image](https://user-images.githubusercontent.com/672429/217152614-fb21967a-f8dc-4d78-a456-9bd41f36383d.png)

The `Distance` struct looks something like 

    public struct Distance
    {
        public Distance(double value, some other args) { // Main constructor.  All other constructors refer to here. }
        public static implicit operator Distance?(string str) => FromString(str); // I thought this would be called from Xaml
        public static Distance? FromString(string str, string unithint = "mm") { // Returns possibly null Distance from string }
        public static Distance Parse(string str) => FromString(str) is Distance d ? d: new Distance(-99); // I think Xaml calls this, can't be null
    }

What constraints must the custom property satisfy so I can bind to it in xaml?  I'm thinking the custom type needs to not be nullable, and have the `Parse` method implemented.  I thought I'd check the experts before rewriting everything to be non-nullable.

thank you

**Comment by sonicsmooth** [comment]
ok I answered my own question -- looks like the property backing type needs to be non-nullable, and the `Parse` needs to be implemented.

**Reply by timunie** [reply]
really? nullable should work normally, I think. Maybe Parse needs to be nullable as well?

**Reply by sonicsmooth** [reply]
After I posted yesterday I realized I had the `Value` property set as private, so it was crashing on Set_Value.  So maybe changing from `Distance?` to `Distance` was not needed. I can check again today to make sure.
